### PR TITLE
Put `ExecutionContext` inside the Tx context

### PR DESCRIPTION
**Pull request by mamcx**
# Description of Changes

As a preparatory step for further PRs, like those related to the new `Query Engine`, most usages of `ExecutionContext` as an additional context in database calls are removed and put inside the `Tx` contexts instead.

# API and ABI breaking changes

* This pr touches everything related to the db
* It could change the metrics. There were many places where the `ExecutionContext` was overwritten and now it should provide more exact measurements.
* Is necessary to check if it passes correctly the `Identity` & `Address` of replicas

# Expected complexity level and risk

3

# Testing

*Describe any testing you've done, and any testing you'd like your reviewers to do,
so that you're confident that all the changes work as expected!*

- [x] *Write a test you've completed here.*
- [ ] *Write a test you want a reviewer to do here, so they can check it off when they're satisfied.*

### CODE REVIEW

File: crates/core/src/execution_context.rs
Comment (mamcx) [inline crates/core/src/execution_context.rs:171]: This variant `ForTests` is only here because the old `ExecutionContext::default()` builds a `Address::Zero`.  I think this should be removed and change to `Internal`, but keep this here in case somebody think is usefull

File: crates/core/src/host/module_host.rs
Comment (mamcx) [inline crates/core/src/host/module_host.rs:770]: I add this here thinking is more correct

File: crates/core/src/host/wasm_common/module_host_actor.rs
Comment (mamcx) [inline crates/core/src/host/wasm_common/module_host_actor.rs:446]: I need to know how deal with this `ctx`

File: crates/core/src/host/wasmtime/wasm_instance_env.rs
Comment (mamcx) [inline crates/core/src/host/wasmtime/wasm_instance_env.rs:689]: Also this one

File: crates/core/src/subscription/module_subscription_actor.rs
Comment (mamcx) [inline crates/core/src/subscription/module_subscription_actor.rs:211]: Is this correct?

File: crates/core/src/host/module_host.rs
Comment (mamcx) [inline crates/core/src/host/module_host.rs:763]: I add this here thinking is more correct

Comment (mamcx) [reply crates/core/src/host/module_host.rs:763]: This pr fails `test_tx_inputs_are_in_the_commitlog`. Is because:

```rust
 fn do_durability(durability: &dyn Durability<TxData = Txdata>, ctx: &ExecutionContext, tx_data: &TxData) {
   ....
        } else {
...
            debug_assert!(!matches!(
                ctx.reducer_context().map(|rcx| rcx.name.strip_prefix("__identity_")),
                Some(Some("connected__" | "disconnected__"))
            ));
        }
    }```

So what does this mean here? @gefjon

**Review by cloutiertyler — COMMENTED** [review]
Just want to comment that I am good to approve the change to the traits when this is ready.

File: crates/core/src/execution_context.rs
Comment (joshua-spacetime) [reply crates/core/src/execution_context.rs:171]: Can you `#[cfg(test)]` it?

File: crates/core/src/host/wasm_common/module_host_actor.rs
Comment (joshua-spacetime) [reply crates/core/src/host/wasm_common/module_host_actor.rs:462]: We should just remove it since it's already embedded inside of `tx`.

File: crates/core/src/host/wasmtime/wasm_instance_env.rs
Comment (joshua-spacetime) [reply crates/core/src/host/wasmtime/wasm_instance_env.rs:689]: Same here. `env` holds the `tx` which in turn holds the execution context. We're not explicitly passing it to `insert` anymore, so we can just remove this call.

File: crates/core/src/subscription/module_subscription_actor.rs
Comment (joshua-spacetime) [reply crates/core/src/subscription/module_subscription_actor.rs:211]: This comment is no longer applicable after your change, so you can remove it.

File: crates/core/src/subscription/module_subscription_actor.rs
Comment (joshua-spacetime) [inline crates/core/src/subscription/module_subscription_actor.rs:216]: This entire `if let` block can be removed since `read_tx` produced by `commit_tx_downgrade` automatically contains the correct execution context.

File: crates/core/src/host/module_host.rs
Comment (joshua-spacetime) [reply crates/core/src/host/module_host.rs:763]: I think the issue is
```rust
// scheduled reducer name not fetched yet, anyway this is only for logging purpose
const REDUCER: &str = "scheduled_reducer";
```

Notice that `REDUCER` isn't the name of the reducer being scheduled.

I think what you want is the result of `call_reducer_params()` which returns a
```rust
CallReducerParams{ reducer_id, .. }
```

And then `self.info.reducers_map.lookup_name(reducer_id)` will give you the name of the reducer.

File: crates/core/src/db/datastore/locking_tx_datastore/tx.rs
Comment (joshua-spacetime) [inline crates/core/src/db/datastore/locking_tx_datastore/tx.rs:64]: Note: `record_metrics` should take ownership of the `ctx`

File: crates/core/src/db/datastore/locking_tx_datastore/mut_tx.rs
Comment (joshua-spacetime) [inline crates/core/src/db/datastore/locking_tx_datastore/mut_tx.rs:1061]: Note, downgrading is the only way to switch contexts. In particular we downgrade a reducer `tx` in order to execute its corresponding incremental update `tx`. This means we should create a new `ctx` here instead of passing the old one.

File: crates/core/src/db/datastore/locking_tx_datastore/tx.rs
Comment (mamcx) [reply crates/core/src/db/datastore/locking_tx_datastore/tx.rs:64]: Do you want to change it to 

```rust
pub(super) fn record_metrics(
    ctx: ExecutionContext,
..
)```?

File: crates/core/src/db/datastore/locking_tx_datastore/mut_tx.rs
Comment (mamcx) [reply crates/core/src/db/datastore/locking_tx_datastore/mut_tx.rs:1061]: What I should create here? Or you mean I need to pass it on ` pub fn commit_downgrade(self, new_ctx:...)`

Comment (mamcx) [reply crates/core/src/db/datastore/locking_tx_datastore/mut_tx.rs:1061]: Ok done with `(commit | rollback)_mut_tx_downgrade(tx, Workload::Subscribe)`

File: crates/core/src/host/module_host.rs
Comment (mamcx) [reply crates/core/src/host/module_host.rs:763]: Note: This comment is not for the failing test but how construct  the context of `REDUCER`

File: crates/core/src/db/datastore/locking_tx_datastore/mut_tx.rs
Comment (joshua-spacetime) [reply crates/core/src/db/datastore/locking_tx_datastore/mut_tx.rs:1061]: We decided on passing the workload to `commit_downgrade`

File: crates/core/src/db/datastore/locking_tx_datastore/mut_tx.rs
Comment (mamcx) [inline crates/core/src/db/datastore/locking_tx_datastore/mut_tx.rs:1079]: Same here

File: crates/core/src/db/datastore/locking_tx_datastore/tx.rs
Comment (joshua-spacetime) [inline crates/core/src/db/datastore/locking_tx_datastore/tx.rs:49]: Just pointing out that this is dead code too.

File: crates/core/src/host/module_host.rs
Comment (joshua-spacetime) [inline crates/core/src/host/module_host.rs:762]: Could we pass `Workload::Reducer`, but have the reducer context be `None` until we fetch the name?

File: crates/core/src/host/scheduler.rs
Comment (joshua-spacetime) [inline crates/core/src/host/scheduler.rs:90]: Is this init?

File: crates/core/src/host/wasm_common/module_host_actor.rs
Comment (joshua-spacetime) [inline crates/core/src/host/wasm_common/module_host_actor.rs:462]: We don't need `ctx` or the `TODO` anymore, right?

File: crates/core/src/subscription/module_subscription_actor.rs
Comment (joshua-spacetime) [inline crates/core/src/subscription/module_subscription_actor.rs:195]: Note: This should be `Workload::Update` for incremental update. `Workload::Subscribe` is for the initial query.

File: crates/core/src/subscription/module_subscription_actor.rs
Comment (joshua-spacetime) [inline crates/core/src/subscription/module_subscription_actor.rs:202]: Same here.

File: crates/core/src/db/relational_db.rs
Comment (joshua-spacetime) [inline crates/core/src/db/relational_db.rs:605]: Technically `do_durability` only needs a `ReducerContext`. Could you update its function signature? That way we only need to clone the `ReducerContext` here instead of the `ExecutionContext`.

File: crates/core/src/db/datastore/locking_tx_datastore/tx.rs
Comment (mamcx) [reply crates/core/src/db/datastore/locking_tx_datastore/tx.rs]: Done

File: crates/core/src/host/scheduler.rs
Comment (mamcx) [reply crates/core/src/host/scheduler.rs:90]: I don't follow. What you expect here?

**Comment by joshua-spacetime** [comment]
@cloutiertyler this is ready for your approval

File: crates/core/src/db/datastore/locking_tx_datastore/committed_state.rs
Comment (mamcx) [reply crates/core/src/db/datastore/locking_tx_datastore/committed_state.rs:57]: Removed

File: crates/core/src/db/datastore/locking_tx_datastore/tx.rs
Comment (mamcx) [reply crates/core/src/db/datastore/locking_tx_datastore/tx.rs:49]: Done

File: crates/core/src/db/relational_db.rs
Comment (mamcx) [reply crates/core/src/db/relational_db.rs:605]: Changed.

File: crates/core/src/host/module_host.rs
Comment (mamcx) [reply crates/core/src/host/module_host.rs:762]: Will end up as before this suggestion.

File: crates/core/src/host/scheduler.rs
Comment (joshua-spacetime) [reply crates/core/src/host/scheduler.rs:90]: I just want to make sure `Workload::Internal` is the correct workload type. Is this called during init?

**Comment by mamcx** [comment]
> @cloutiertyler this is ready for your approval

Pls wait until I merge https://github.com/clockworklabs/SpacetimeDB/pull/1880 so it will be easier to read...

File: crates/core/src/db/datastore/locking_tx_datastore/mut_tx.rs
Comment (joshua-spacetime) [inline crates/core/src/db/datastore/locking_tx_datastore/mut_tx.rs:1444]: Oops, more dead code :)

File: crates/core/src/db/relational_db.rs
Comment (joshua-spacetime) [inline crates/core/src/db/relational_db.rs:620]: Does this need to be cloned? You pass it by reference to `do_durability`.

File: crates/core/src/host/wasm_common/module_host_actor.rs
Comment (joshua-spacetime) [inline crates/core/src/host/wasm_common/module_host_actor.rs:459]: This still hasn't been removed.

File: crates/core/src/db/relational_db.rs
Comment (mamcx) [reply crates/core/src/db/relational_db.rs:620]: Yes:

```rust
        let reducer_context = tx.ctx.reducer_context();
        let Some(tx_data) = self.inner.commit_mut_tx(tx)? else {
            return Ok(None);
        };
```

File: crates/core/src/host/wasm_common/module_host_actor.rs
Comment (joshua-spacetime) [reply crates/core/src/host/wasm_common/module_host_actor.rs:459]: We can create a follow up ticket for this one.

**Review by joshua-spacetime — APPROVED** [review]

File: crates/core/src/host/wasm_common/module_host_actor.rs
Comment (mamcx) [reply crates/core/src/host/wasm_common/module_host_actor.rs:459]: Context: Is not possible to remove it without causing a deadlock, so is better to wait for a proper investigation.

**Review by cloutiertyler — APPROVED** [review]
LGTM

**Pull request merged**
